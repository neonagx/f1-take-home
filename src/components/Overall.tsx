import React, { FC, useEffect, useState } from "react";
import { Link } from "react-router-dom";
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Button from '@mui/material/Button';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';

interface Driver {
    dateOfBirth: string;
    driverId: string;
    givenName: string;
    familyName: string;
    nationality: string;
}

interface OverallData {
    position: string;
    Driver: Driver
}

interface OverallRaces {
    round: string;
}

const Overall: FC = () => {
    const [overallData, setOverallData] = useState<OverallData[]>([]);
    const [overallRaces, setOverallRaces] = useState<OverallRaces[]>([]);

    const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
    const open = Boolean(anchorEl);

    const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
        setAnchorEl(event.currentTarget);
    };
    const handleClose = () => {
        setAnchorEl(null);
    };

    useEffect(() => {
        const getOverallData = async () => {
            await fetch('http://ergast.com/api/f1/current/driverStandings.json')
                .then(res => res.json())
                .then(results => setOverallData(results['MRData']['StandingsTable']['StandingsLists'][0]['DriverStandings']))
                .catch(err => console.log(err))
        };

        const getOverallRaces = async () => {
            await fetch('https://ergast.com/api/f1/current/races.json')
                .then(res => res.json())
                .then(results => setOverallRaces(results['MRData']['RaceTable']['Races']))
                .catch(err => console.log(err))
        };
        getOverallData()
        getOverallRaces();
    }, []);

    return (
        <div>
            <h2>Overall Standing 2022</h2>

            <div>
                <Button
                    id="basic-button"
                    aria-controls={open ? 'basic-menu' : undefined}
                    aria-haspopup="true"
                    aria-expanded={open ? 'true' : undefined}
                    onClick={handleClick}
                >
                    Go To Race Results
                </Button>
                <Menu
                    className="basic-menu"
                    anchorEl={anchorEl}
                    open={open}
                    onClose={handleClose}
                    MenuListProps={{
                        'aria-labelledby': 'basic-button',
                    }}
                >
                    {overallRaces.map((race) => <MenuItem key={race['round']} onClick={handleClose}><Link to={`/raceresults/${race['round']}`}>{race['round']}</Link></MenuItem>)}
                </Menu>
            </div>

            <TableContainer>
                <Table sx={{ minWidth: 500 }} aria-label="simple table">
                    <TableHead>
                        <TableRow>
                            <TableCell align="left">Position</TableCell>
                            <TableCell align="left">Driver</TableCell>
                            <TableCell align="left">Date Of Birth</TableCell>
                            <TableCell align="left">Nationality</TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {overallData.map(data => (
                            <TableRow
                                key={data['position']}
                                sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
                                <TableCell align="left">{data['position']}</TableCell>
                                <TableCell align="left">{<Link to={`/driverresults/${data['Driver']['driverId']}`}>{`${data['Driver']['givenName']} ${data['Driver']['familyName']}`}</Link>}</TableCell>
                                <TableCell align="left">{data['Driver']['dateOfBirth']}</TableCell>
                                <TableCell align="left">{data['Driver']['nationality']}</TableCell>
                            </TableRow>
                        ))}
                    </TableBody>
                </Table>
            </TableContainer>
        </div>

    );
}

export default Overall;