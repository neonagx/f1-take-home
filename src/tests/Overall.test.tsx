import React from 'react';
import { render, screen } from '@testing-library/react';
import Enzyme, { shallow } from 'enzyme';
import Overall from '../components/Overall';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });
React.useLayoutEffect = React.useEffect;

describe('<Overall />', () => {

    test('should have Overall Standing Page', () => {
        const wrapper = shallow(<Overall />);

        expect(wrapper.html()).toMatchSnapshot();
    });


})
