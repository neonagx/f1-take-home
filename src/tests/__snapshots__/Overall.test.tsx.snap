// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overall /> should have Overall Standing Page 1`] = `"<div><h2>Overall Standing 2022</h2><div><button class=\\"MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root\\" tabindex=\\"0\\" type=\\"button\\" id=\\"basic-button\\" aria-haspopup=\\"true\\">Go To Race Results</button></div><div class=\\"MuiTableContainer-root css-rorn0c-MuiTableContainer-root\\"><table class=\\"MuiTable-root css-130xbax-MuiTable-root\\" aria-label=\\"simple table\\"><thead class=\\"MuiTableHead-root css-15wwp11-MuiTableHead-root\\"><tr class=\\"MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root\\"><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root\\" scope=\\"col\\">Position</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root\\" scope=\\"col\\">Driver</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root\\" scope=\\"col\\">Date Of Birth</th><th class=\\"MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root\\" scope=\\"col\\">Nationality</th></tr></thead><tbody class=\\"MuiTableBody-root css-apqrd9-MuiTableBody-root\\"></tbody></table></div></div>"`;
